File: python/conference_test.py
import unittest
from python import conference


class ConferenceRoomSystemTests(unittest.TestCase):

    def setUp(self):
        self.crs = conference.ConferenceRoomSystem()

    def test_add_building(self):
        self.crs.add_building("b1")
        self.assertEqual(len(self.crs.Buildings), 1)

        self.crs.add_building("b2")
        self.assertEqual(len(self.crs.Buildings), 2)

        self.crs.add_building("b1")
        self.assertEqual(len(self.crs.Buildings), 2)

    def test_add_floor(self):
        self.crs.add_building("b1")
        self.crs.add_floor("b1", 1)
        self.assertEqual(len(self.crs.Buildings["b1"].Floors), 1)

        self.crs.add_floor("b1", 2)
        self.assertEqual(len(self.crs.Buildings["b1"].Floors), 2)

        self.crs.add_floor("b2", 1)

    def test_add_conference_room(self):
        self.crs.add_building("b1")
        self.crs.add_floor("b1", 1)
        self.crs.add_conference_room("b1", 1, "c1")
        self.assertEqual(len(self.crs.Buildings["b1"].Floors[1]), 1)

        self.crs.add_conference_room("b1", 1, "c1")

        self.crs.add_conference_room("b2", 1, "c2")

        self.crs.add_conference_room("b1", 2, "c2")

    def test_book_conference_room(self):
        self.crs.add_building("b1")
        self.crs.add_floor("b1", 1)
        self.crs.add_conference_room("b1", 1, "c1")

        self.crs.book_conference_room("1:3", "b1", 1, "c1")
        self.assertEqual(len(self.crs.Buildings["b1"].ConferenceMap["c1"].Bookings), 1)

        self.crs.book_conference_room("1:3", "b1", 1, "c2")

        self.crs.book_conference_room("1:3", "b2", 1, "c1")

    def test_cancel_booking(self):
        self.crs.add_building("b1")
        self.crs.add_floor("b1", 1)
        self.crs.add_conference_room("b1", 1, "c1")
        self.crs.book_conference_room("1:3", "b1", 1, "c1")

        self.crs.cancel_booking("1:3", "b1", 1, "c1")
        self.assertEqual(len(self.crs.Buildings["b1"].ConferenceMap["c1"].Bookings), 0)

        self.crs.cancel_booking("1:3", "b1", 1, "c2")

        self.crs.cancel_booking("1:3", "b2", 1, "c1")

    def test_list_rooms(self):
        self.crs.add_building("b1")
        self.crs.add_floor("b1", 1)
        self.crs.add_conference_room("b1", 1, "c1")
        self.crs.add_building("b2")
        self.crs.add_floor("b2", 2)
        self.crs.add_conference_room("b2", 2, "c2")

        expected_output = None
        self.assertEqual(self.crs.list_rooms(), expected_output)

    def test_list_bookings(self):
        self.crs.add_building("b1")
        self.crs.add_floor("b1", 1)
        self.crs.add_conference_room("b1", 1, "c1")
        self.crs.book_conference_room("1:3", "b1", 1, "c1")

        expected_output = None
        self.assertEqual(self.crs.list_bookings("b1", 1), expected_output)


if __name__ == '__main__':
    unittest.main()

File: golang/conference_test.go
package main

import (
	"bytes"
	"io"
	"os"
	"testing"
)

func TestConferenceRoomSystem(t *testing.T) {
	crs := NewConferenceRoomSystem()

	// Test AddBuilding
	crs.AddBuilding("b1")
	if _, ok := crs.Buildings["b1"]; !ok {
		t.Errorf("AddBuilding failed, building 'b1' not found in the system")
	}

	// Test AddFloor
	err := crs.AddFloor("b1", 1)
	if err != nil {
		t.Errorf("AddFloor failed: %v", err)
	}

	// Test AddConferenceRoom
	err = crs.AddConferenceRoom("b1", 1, "c1")
	if err != nil {
		t.Errorf("AddConferenceRoom failed: %v", err)
	}

	// Test BookConferenceRoom
	err = crs.BookConferenceRoom("1:3", "b1", 1, "c1")
	if err != nil {
		t.Errorf("BookConferenceRoom failed: %v", err)
	}

	// Test ListRooms
	expectedListRoomsOutput := "c1 1 b1 []\n"
	listRoomsOutput := captureOutput(func() {
		crs.ListRooms()
	})
	if listRoomsOutput != expectedListRoomsOutput {
		t.Errorf("ListRooms failed, expected output: '%s', got: '%s'", expectedListRoomsOutput, listRoomsOutput)
	}

	// Test ListBookings
	expectedListBookingsOutput := ""
	listBookingsOutput := captureOutput(func() {
		crs.ListBookings("b1", 1)
	})
	if listBookingsOutput != expectedListBookingsOutput {
		t.Errorf("ListBookings failed, expected output: '%s', got: '%s'", expectedListBookingsOutput,
			listBookingsOutput)
	}

	// Test CancelBooking
	err = crs.CancelBooking("1:3", "b1", 1, "c1")
	if err != nil {
		t.Errorf("CancelBooking failed: %v", err)
	}

	// Test ListRooms after cancellation
	expectedListRoomsOutputAfterCancellation := "c1 1 b1 []\n"
	listRoomsOutputAfterCancellation := captureOutput(func() {
		crs.ListRooms()
	})
	if listRoomsOutputAfterCancellation != expectedListRoomsOutputAfterCancellation {
		t.Errorf("ListRooms after cancellation failed, expected output: '%s', got: '%s'",
			expectedListRoomsOutputAfterCancellation, listRoomsOutputAfterCancellation)
	}
}

// Helper function to capture stdout output
func captureOutput(f func()) string {
	old := os.Stdout
	r, w, _ := os.Pipe()
	os.Stdout = w

	f()

	w.Close()
	os.Stdout = old

	var buf bytes.Buffer
	io.Copy(&buf, r)

	return buf.String()
}
